File: 1.Fundamental-Level/1.Advanced-C#/00.Exam-Preparation/Exam-Problems-Practice/11.LittleJohn/LittleJohn.cs
namespace _11.LittleJohn
{
    using System;
    using System.Linq;
    using System.Text;
    using System.Text.RegularExpressions;

    public class LittleJohn
    {
        public static void Main()
        {
            StringBuilder arrowsBuilder = new StringBuilder();
            for (int i = 0; i < 4; i++)
            {
                arrowsBuilder.AppendFormat(" {0}", Console.ReadLine());
            }

            const string ArrowPattern = @"(>>>----->>)|(>>----->)|(>----->)";
            Regex arrowMatcher = new Regex(ArrowPattern);
            var arrows = arrowMatcher.Matches(arrowsBuilder.ToString());

            int largeArrowCount = 0;
            int mediumArrowCount = 0;
            int smallArrowCount = 0;

            foreach (Match arrow in arrows)
            {
                if (!string.IsNullOrEmpty(arrow.Groups[1].Value))
                {
                    largeArrowCount++;
                }
                else if (!string.IsNullOrEmpty(arrow.Groups[2].Value))
                {
                    mediumArrowCount++;
                }
                else if (!string.IsNullOrEmpty(arrow.Groups[3].Value))
                {
                    smallArrowCount++;
                }
            }

            string numberAsString = string.Format("{0}{1}{2}", smallArrowCount, mediumArrowCount, largeArrowCount);

            long number = long.Parse(numberAsString);

            string binaryNum = Convert.ToString(number, 2);

            string reversedBinaryNum = new string(binaryNum.Reverse().ToArray());

            string resultAsString = binaryNum + reversedBinaryNum;

            long result = Convert.ToInt64(resultAsString, 2);

            Console.WriteLine(result);
        }
    }
}

File: Java-Development/Advanced-Java/08.Stream-Api-Exercises/src/_12_LittleJohn/LittleJohn.java
package _12_LittleJohn;

import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class LittleJohn {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        String arrows = "";
        for (int i = 0; i < 4; i++) {
            arrows += " " + scanner.nextLine();
        }

        Pattern pattern = Pattern.compile("(>>>----->>)|(>>----->)|(>----->)");
        Matcher matcher = pattern.matcher(arrows);

        int largeArrowCount = 0;
        int mediumArrowCount = 0;
        int smallArrowCount = 0;
        while (matcher.find()) {
            if (matcher.group(1) != null) {
                largeArrowCount++;
            }

            if (matcher.group(2) != null) {
                mediumArrowCount++;
            }

            if (matcher.group(3) != null) {
                smallArrowCount++;
            }
        }

        String numberAsString = smallArrowCount + "" + mediumArrowCount + largeArrowCount;
        Long number = Long.parseLong(numberAsString);

        String binaryNum = Long.toBinaryString(number);

        String reversedBinaryNum = new StringBuilder(binaryNum).reverse().toString();

        String resultAsString = binaryNum + reversedBinaryNum;

        long result = Long.parseLong(resultAsString, 2);

        System.out.println(result);
    }
}
